# Lista 1 - Econometria II
---

> Alunos: Henrique Castilho Del'Esposti, Luciana Thomasi, Paola Scheidegger e Renan Rodrigues Torres.

# Questão 1

## a) 


A comparação de médias entre os grupos inicialmente propostos não é adequeda por dois motivos: i) ambos são compostos por indivíduos distintos e com características que podem se manifestar significativamente diferentes do ponto de vista estatístico; ii) como narrado, os indivíduos aprovados no processo seletivo para ingressar nas forças armadas passam por avaliações criteriosas que avaliam características que também estão ou podem estar relacionadas com os salários no futuro, como escolaridade, notas em exames e aptidão física. Desse modo, essa simples comparação desconsidera esses fatores que impactam os salários, além da própria hipótese do impacto do serviço militar nos salários.

## b) 

O grupo de tratamento ideal seriam os próprios candidatos aprovados para ingressar nas forças armadas. Enquanto isso, um grupo de controle ideal seria composto por indivíduos que possuem características similares em termos de escolaridade, idades, notas dos exames e aptidão física, mas que por algum motivo aleatório nã ingressaram nas forças armadas.

Em aplicações práticas, os grupos ideais podem possuir maior flexibilidade e terem as médias de atributos distintas, dado que o que realmente importa é que, no tempo, a variação no grupo de controle antes e depois do programa reflita uma variação no tempo semelhante ao grupo dos tratados em uma hipótese contrafactual, isto é, caso não fossem tratados.

## c) 

A regressão para o modelo de Diferenças-em-Diferenças a ser estimada seria, seguindo a conveção apresentada nos livros-texto:

$$
Y_{it} = \alpha_0 + \alpha_1educ + \alpha_2idade + \alpha_3notas + \alpha_4saude + \gamma_1T_i + \varrho_1dt_t + \beta_1D_{it} + \varepsilon_{it}
$$

- educ: variável para o grau de escolaridade. Sua inclusão justifica-se pelo fato de impor controle sobre os efeitos da educação em relação à variável de interesse Y, o salário, dados os efeitos positivos que o nível educacional pode exercer sobre os salários, ou seja, ao deixarmos de fora, poderíamos estar superestimando o parâmetro de interesse $\beta_1$.

- idade: variável para a idade dos indivíduos analisados. Sua inclusão justifica-se pelo fato de impor controle sobre os efeitos da idade em relação à variável de interesse Y, o salário. Nesse caso, com a maior idade os indivíduos podem adquirir mais experiência no mercado de trabalho, o que poderia influenciar em maiores níveis salariais, isto é, poderíamos estar superestimando o parâmetro de interesse $\beta_1$ caso não controlássemos para essa variável.

- notas: variável para a nota obtida nos exames teóricos para ingressar nas força armadas. Sua inclusão justifica-se pelo fato de impor controle sobre os efeitos das notas em relação à variável de interesse Y, o salário. Isso porque a nota pode refletir parcialmente a capacidade dos indivíduos em ingressarem nas forças armadas, de modo que maiores notas pudessem representar uma maior probabilidade de ingresso nas forças armadas. Sem controlar para as notas, poderíamos estar superestimando o parâmetro de interesse $\beta_1$ caso não controlássemos para essa variável.

- saude: variável binária que assumiria valor unitário para a presença de doenças crônicas que afetam a aptidão física dos indivíduos e zero caso contrário. Sua inclusão justifica-se pelas avaliações físicas dos candidatos ao exame das forças armadas. Sem controlar para esses efeitos, poderíamos estar superestimando o parâmetro de interesse $\beta_1$ caso não controlássemos para essa variável.

- $D_{it}$: interação entre as dummies $T_i$ e $dt$ que identificar a presença de tratamento dos indivíduos e o período pós-intervenção, que assumiria o valor 1 caso o indivíduo da amostra recebesse o tratamento. Nesse caso, a inclusão dessa interação serve para estimar o parâmetro de interesse $\beta_1$ para capturar o impacto da adesão das forças armadas sobre os salários.

# Questão 2

## a) 

Verdadeiro. Dado que temos variáveis explicativas exógenas, isto é, sem estarem correlacionadas com o termo do erro, basta termos uma amostras significativamente grande para que o valor estimado do parâmetro seja convergente com o seu valor na população. 

## b) 

Falso. Com a presença de multicolinearidade, temos que os erros padrões dos nossos estimadores tendem a ser consideravelmente elevados, o que inviabiliza os testes de hipótese e a inferência, mas não implicam na presença de viés em relação aos estimadores. 

## c) 

Falso. Por estar relacionada com a variância dos erros, a hipótese de homoscedasticidade apenas viabiliza os testes de hipótese e a inferência sobre a significância dos parâmetros.  

# Questão 3)

## a) 

Substituindo $y_1$ em $y_2$:

$$
y_2 = \gamma_0 + \gamma_1(\phi_0 + \phi_1y_2 + \phi_2x_1 + \phi_3x_2 + \varepsilon) + \gamma_2x_1 + \gamma_3x_3 + \epsilon
$$

Aplicando a distributiva, retirando os interceptos para didática e manipulando algebricamente:

$$
(1-\gamma_1\phi_1)y_2 = \gamma_1\phi_2x_1 + \gamma_1\phi_3x_2 + \gamma_1\varepsilon + \gamma_2x_1 + \gamma_3x_3 + \epsilon
$$ 

$$
y_2 = \frac{\gamma_1\phi_2x_1}{(1-\gamma_1\phi_1)} + \frac{\gamma_1\phi_3x_2}{(1-\gamma_1\phi_1)} + \frac{\gamma_2x_1}{(1-\gamma_1\phi_1)} + \frac{\gamma_3x_3}{(1-\gamma_1\phi_1)} + \frac{\gamma_1\varepsilon + \epsilon}{(1-\gamma_1\phi_1)}
$$

Com $\gamma_1\phi_1 \neq 0$ e $π_{12} = \frac{\gamma_1\phi_2}{(1-\gamma_1\phi_1)} + \frac{\gamma_2x_1}{(1-\gamma_1\phi_1)}$, $π_{13} = \frac{\gamma_1\phi_3}{(1-\gamma_1\phi_1)}$, $\pi_3 = \frac{\gamma_3}{(1-\gamma_1\phi_1)}$ e $v_2 = \frac{\gamma_1\varepsilon + \epsilon}{(1-\gamma_1\phi_1)}$, teremos:

$$
y_2 = \pi_{12}x_1 + \pi_{13}x_2 + \pi_3x_3 + v_2 
$$


## b)

Não, para que a equação (1) possa ser identificada, é necessário que $\gamma_3 \neq 0$, dado que esse coeficiente está ligado a variável $x_3$, que não aparece na equação (1) e se torna o critério para identificá-la. 

## c)

Sim. Se $\phi_1 = 0$, a variável $y_2$ não é estatisticamente significante para explicar $y_1$ e, assim, não teremos o problema do viés de simultaneidade. Dessa forma, se $x_1$ e $x_2$ forem exógenas, podemos estimar a equação (1) por MQO. 

## d)

Sim. Com essas condições, teremos a possibilidade de identificar a equação (1) e também sabemos que estimá-la por MQO será equivocado por causa do viés de simultaneidade. Esse viés opera de forma similar ao problema da endogeneidade, de modo que a variável explicada simultaneamente por outra estará correlacionada com o termo do erro, fazendo com que os estimadores sejam incosistentes.

# *Observação*

Antes de solucionar as questões com os dados, carregamos e convertemos os arquivos .RData para .xlsx (Excel) para trabalhar no Python, com o seguinte código em R:

***install.packages("writexl")***

***library("writexl")***

***write_xlsx(data, "microcredito_es.xlsx")***

***write_xlsx(dados, "mercado_peixe.xlsx")***


# Questão 4) 

In [48]:
import pandas as pd
import numpy as np

In [49]:
microcred_esdata = pd.read_excel('microcredito_es.xlsx')
microcred_esdata

municipio   ano      cred_def       FPM_def  royalties_def  \
0    Afonso Cláudio  2010  2.312024e+06  2.063070e+07   1.555022e+06   
1    Afonso Cláudio  2011  2.481674e+06  2.347283e+07   2.908329e+06   
2    Afonso Cláudio  2012  2.185640e+06  2.282401e+07   3.398179e+06   
3    Afonso Cláudio  2013  1.818640e+06  2.317639e+07   3.507842e+06   
4    Afonso Cláudio  2014  1.882815e+06  2.325341e+07   3.816122e+06   
..              ...   ...           ...           ...            ...   
541         Vitória  2012  5.792563e+06  1.510206e+08   2.555672e+07   
542         Vitória  2013  3.946719e+06  1.357670e+08   2.981084e+07   
543         Vitória  2014  4.476135e+06  1.411357e+08   3.388286e+07   
544         Vitória  2015  3.886005e+06  1.348078e+08   2.217259e+07   
545         Vitória  2016  3.148692e+06  1.666094e+08   1.717894e+07   

          pib_def  oper  IFDM_GERAL  esti_receitapropria     pop  
0    4.050472e+08   193    0.712163         3.110818e+06   31091  
1    4.338577e+08   212    0.737355         3.307073e+06   31004  
2    4.753806e+08   199    0.748059         3.444694e+06   30919  
3    4.810916e+08   193    0.736154         3.609006e+06   32551  
4    4.917057e+08   179    0.708248         4.014598e+06   32502  
..            ...   ...         ...                  ...     ...  
541  3.586177e+10   670    0.851657         7.024802e+08  333162  
542  3.097523e+10   418    0.854752         6.534731e+08  348268  
543  3.066487e+10   677    0.857995         7.102522e+08  352104  
544  2.726150e+10   352    0.831387         6.217284e+08  355875  
545  2.415886e+10   250    0.824399         5.725888e+08  359555  

[546 rows x 10 columns]

## a)

In [50]:
microcred_esdata['cred_def_pc'] = microcred_esdata['cred_def'] / microcred_esdata['pop']
microcred_esdata['FPM_def_pc'] = microcred_esdata['FPM_def'] / microcred_esdata['pop']
microcred_esdata['royalties_def_pc'] = microcred_esdata['royalties_def'] / microcred_esdata['pop']
microcred_esdata['pib_def_pc'] = microcred_esdata['pib_def'] / microcred_esdata['pop']
microcred_esdata['esti_receitapropria_pc'] = microcred_esdata['esti_receitapropria'] / microcred_esdata['pop']

microcred_esdata_pc = microcred_esdata[['cred_def_pc', 'FPM_def_pc', 'royalties_def_pc', 'pib_def_pc', 'esti_receitapropria_pc']]
microcred_esdata_pc

cred_def_pc  FPM_def_pc  royalties_def_pc     pib_def_pc  \
0      74.363115  663.558598         50.015186   13027.795730   
1      80.043674  757.090288         93.804959   13993.602539   
2      70.689225  738.187337        109.905849   15375.032463   
3      55.870489  712.002355        107.764494   14779.626804   
4      57.929191  715.445407        117.411904   15128.474183   
..           ...         ...               ...            ...   
541    17.386625  453.294809         76.709601  107640.639000   
542    11.332420  389.834861         85.597408   88940.786253   
543    12.712537  400.835313         96.229701   87090.380672   
544    10.919579  378.806538         62.304438   76604.147478   
545     8.757191  463.376776         47.778335   67191.007749   

     esti_receitapropria_pc  
0                100.055246  
1                106.666000  
2                111.410271  
3                110.872349  
4                123.518499  
..                      ...  
541             2108.524305  
542             1876.351285  
543             2017.165855  
544             1747.041558  
545             1592.493070  

[546 rows x 5 columns]

## b) 

In [51]:
import statsmodels.formula.api as smf
import statsmodels.api as sm

In [52]:
reg_microes_ols = smf.ols('pib_def_pc ~ cred_def_pc', data = microcred_esdata)
results_reg_microes_ols = reg_microes_ols.fit()
print(results_reg_microes_ols.summary())

                            OLS Regression Results                            
Dep. Variable:             pib_def_pc   R-squared:                       0.000
Model:                            OLS   Adj. R-squared:                 -0.002
Method:                 Least Squares   F-statistic:                    0.1294
Date:                Fri, 15 Jul 2022   Prob (F-statistic):              0.719
Time:                        11:36:14   Log-Likelihood:                -7060.2
No. Observations:                 546   AIC:                         1.412e+04
Df Residuals:                     544   BIC:                         1.413e+04
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
Intercept    4.052e+04   8146.070      4.974      

O sinal encontrado no coeficiente não está de acordo com o esprado pela teoria econômica: esperava-se que um aumento no microcrédito elevasse a atividade econômica, e não o contrário. 

Contudo, o parâmetro não é significante ao nível de 5%, dado que temos uma probabilidade de 71,9% (p-valor = 0.719) de errarmos a inferência caso rejeitarmos a hipótese nula de que $\beta_1 = 0$.

Do ponto de vista estatístico, vale lembrar que estimar o modelo por MQO e com apenas o microcrédito como variável explicativa nos leva a obter um estimador viesado por ignorar alguns problemas como a endogeneidade, por outras variáveis do termo do erro estarem relacionadas com a variável de interesse e a própria variável explicativa, e também as especificidades não observadas de cada cidade contemplada pelo programa de microcrédito, que podem influenciar nos impactos dessa política.

## c) 

Transformamos a variável *oper* para per capita também, para controlarmos as operações de crédito para o tamanho do município em termos de população.

In [53]:
microcred_esdata['oper_pc'] = microcred_esdata['oper'] / microcred_esdata['pop']
microcred_esdata[['oper_pc']]

oper_pc
0    0.006208
1    0.006838
2    0.006436
3    0.005929
4    0.005507
..        ...
541  0.002011
542  0.001200
543  0.001923
544  0.000989
545  0.000695

[546 rows x 1 columns]

In [54]:
reg_microes_ols_4c = smf.ols('pib_def_pc ~ cred_def_pc + FPM_def_pc + royalties_def_pc + oper_pc + IFDM_GERAL', 
                              data = microcred_esdata)
results_reg_microes_ols_4c = reg_microes_ols_4c.fit()
print(results_reg_microes_ols_4c.summary())

                            OLS Regression Results                            
Dep. Variable:             pib_def_pc   R-squared:                       0.931
Model:                            OLS   Adj. R-squared:                  0.930
Method:                 Least Squares   F-statistic:                     1450.
Date:                Fri, 15 Jul 2022   Prob (F-statistic):          3.44e-310
Time:                        11:36:14   Log-Likelihood:                -6331.7
No. Observations:                 546   AIC:                         1.268e+04
Df Residuals:                     540   BIC:                         1.270e+04
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
Intercept         -2.98e+04    1.7e+04  

## d)

In [55]:
import linearmodels as plm

microcred_esdata_ef = microcred_esdata.copy() 
microcred_esdata_ef.set_index(['municipio', 'ano'], inplace = True)
microcred_esdata_ef

cred_def       FPM_def  royalties_def       pib_def  \
municipio      ano                                                             
Afonso Cláudio 2010  2.312024e+06  2.063070e+07   1.555022e+06  4.050472e+08   
               2011  2.481674e+06  2.347283e+07   2.908329e+06  4.338577e+08   
               2012  2.185640e+06  2.282401e+07   3.398179e+06  4.753806e+08   
               2013  1.818640e+06  2.317639e+07   3.507842e+06  4.810916e+08   
               2014  1.882815e+06  2.325341e+07   3.816122e+06  4.917057e+08   
...                           ...           ...            ...           ...   
Vitória        2012  5.792563e+06  1.510206e+08   2.555672e+07  3.586177e+10   
               2013  3.946719e+06  1.357670e+08   2.981084e+07  3.097523e+10   
               2014  4.476135e+06  1.411357e+08   3.388286e+07  3.066487e+10   
               2015  3.886005e+06  1.348078e+08   2.217259e+07  2.726150e+10   
               2016  3.148692e+06  1.666094e+08   1.717894e+07  2.415886e+10   

                     oper  IFDM_GERAL  esti_receitapropria     pop  \
municipio      ano                                                   
Afonso Cláudio 2010   193    0.712163         3.110818e+06   31091   
               2011   212    0.737355         3.307073e+06   31004   
               2012   199    0.748059         3.444694e+06   30919   
               2013   193    0.736154         3.609006e+06   32551   
               2014   179    0.708248         4.014598e+06   32502   
...                   ...         ...                  ...     ...   
Vitória        2012   670    0.851657         7.024802e+08  333162   
               2013   418    0.854752         6.534731e+08  348268   
               2014   677    0.857995         7.102522e+08  352104   
               2015   352    0.831387         6.217284e+08  355875   
               2016   250    0.824399         5.725888e+08  359555   

                     cred_def_pc  FPM_def_pc  royalties_def_pc     pib_def_pc  \
municipio      ano                                                              
Afonso Cláudio 2010    74.363115  663.558598         50.015186   13027.795730   
               2011    80.043674  757.090288         93.804959   13993.602539   
               2012    70.689225  738.187337        109.905849   15375.032463   
               2013    55.870489  712.002355        107.764494   14779.626804   
               2014    57.929191  715.445407        117.411904   15128.474183   
...                          ...         ...               ...            ...   
Vitória        2012    17.386625  453.294809         76.709601  107640.639000   
               2013    11.332420  389.834861         85.597408   88940.786253   
               2014    12.712537  400.835313         96.229701   87090.380672   
               2015    10.919579  378.806538         62.304438   76604.147478   
               2016     8.757191  463.376776         47.778335   67191.007749   

                     esti_receitapropria_pc   oper_pc  
municipio      ano                                     
Afonso Cláudio 2010              100.055246  0.006208  
               2011              106.666000  0.006838  
               2012              111.410271  0.006436  
               2013              110.872349  0.005929  
               2014              123.518499  0.005507  
...                                     ...       ...  
Vitória        2012             2108.524305  0.002011  
               2013             1876.351285  0.001200  
               2014             2017.165855  0.001923  
               2015             1747.041558  0.000989  
               2016             1592.493070  0.000695  

[546 rows x 14 columns]

In [56]:
reg_microes_ef = plm.PanelOLS.from_formula('pib_def_pc ~ cred_def_pc + FPM_def_pc + royalties_def_pc + oper_pc + IFDM_GERAL +'
                                           'EntityEffects', 
                                           data = microcred_esdata_ef)
results_reg_microes_ef = reg_microes_ef.fit()
print(results_reg_microes_ef.summary)

                          PanelOLS Estimation Summary                           
Dep. Variable:             pib_def_pc   R-squared:                        0.8666
Estimator:                   PanelOLS   R-squared (Between):              0.2522
No. Observations:                 546   R-squared (Within):               0.8666
Date:                Fri, Jul 15 2022   R-squared (Overall):              0.3435
Time:                        11:36:15   Log-likelihood                   -6026.6
Cov. Estimator:            Unadjusted                                           
                                        F-statistic:                      601.80
Entities:                          78   P-value                           0.0000
Avg Obs:                       7.0000   Distribution:                   F(5,463)
Min Obs:                       7.0000                                           
Max Obs:                       7.0000   F-statistic (robust):             601.80
                            

Como a política de microcrédito opera em todos os municípios, devemos controlar para as heterogeneidades não observadas que podem influenciar no impacto do programa. Por ser uma variável importante, inserí-la no termo do erro nso leva ao viés da variável omitida e, por isso, os estimadores obtidos serão viesados. 

Com o modelo de efeitos fixos, controlamos a heterogeneidade não observada dos municípios do termo do erro, obtendo estimadores mais precisos. 

## e) 

In [57]:
reg_microes_eftempo = plm.PanelOLS.from_formula('pib_def_pc ~ cred_def_pc + FPM_def_pc + royalties_def_pc + oper_pc +' 
                                                'IFDM_GERAL + TimeEffects', 
                                                data = microcred_esdata_ef)
results_reg_microes_eftempo = reg_microes_eftempo.fit()
print(results_reg_microes_eftempo.summary)

                          PanelOLS Estimation Summary                           
Dep. Variable:             pib_def_pc   R-squared:                        0.9310
Estimator:                   PanelOLS   R-squared (Between):              0.9007
No. Observations:                 546   R-squared (Within):               0.7426
Date:                Fri, Jul 15 2022   R-squared (Overall):              0.8772
Time:                        11:36:16   Log-likelihood                   -6328.6
Cov. Estimator:            Unadjusted                                           
                                        F-statistic:                      1440.9
Entities:                          78   P-value                           0.0000
Avg Obs:                       7.0000   Distribution:                   F(5,534)
Min Obs:                       7.0000                                           
Max Obs:                       7.0000   F-statistic (robust):             1440.9
                            

In [58]:
results_reg_microes_eftempo.f_pooled

Pooled F-statistic
H0: Effects are zero
Statistic: 1.0145
P-value: 0.4148
Distributed: F(6,534)
WaldTestStatistic, id: 0x1434d6282e0

Ao controlarmos para efeitos fixos de tempo, conseguimos capturar mudanças ocorridas ao longo do tempo de análise e que podem afetar a oferta ou o acesso de crédito, além da própria polítca, podendo então, influenciar em seu impacto. Alterações do custo do crédito com as variações na taxa básica da economia ou mudanças regulatórias no mercado de crédito podem ser alguns exemplos dessas mudanças que podem ser controladas ao adicionarmos os efeitos fixos de tempo na regressão.

Ao compararmos os coeficientes para a variável de micrédito na regressão com efeitos fixos de tempo contra os efeitos fixos para municípios, podemos observar que para o primeiro caso, o sinal não corresponde ao esperado pela teoria econômica. Ainda, de modo individual, vemos que o parâmetro estimado para essa variável torna-se não significante do ponto de vista estatístico, até mesmo ao nível de 10%. 

Por fim, tanto no fim do sumário quanto na linha de código acima, explicitamos os resultados do teste F para a significância dos efeitos de tempo na regressão. Com um p-valor de 0.4148, temos uma probabilidade de 41.48% de errar ao rejeitarmos a hipótese nula de que os efeitos em questão são estatisticamente insignificantes. 

## f)

Podemos definir o modelo teórico para este problema como:

$$
Y_{it} = \beta_0 + \beta_1credpc + \beta_2operpc + \beta_3royaltiespc + \beta_4FPMpc + \beta_5IFDMGERAL + \alpha_i + \lambda_t + u_{it}
$$

para $i = 1, 2, ..., n$ e $t = 1, 2, ..., T$

A seguir, podemos estimá-lo.

In [59]:
reg_microes_eftempo_mun = plm.PanelOLS.from_formula('pib_def_pc ~ cred_def_pc + FPM_def_pc + royalties_def_pc + oper_pc +' 
                                                    'IFDM_GERAL + TimeEffects + EntityEffects', 
                                                    data = microcred_esdata_ef)
results_reg_microes_eftempo_mun = reg_microes_eftempo_mun.fit()
print(results_reg_microes_eftempo_mun.summary)

                          PanelOLS Estimation Summary                           
Dep. Variable:             pib_def_pc   R-squared:                        0.8720
Estimator:                   PanelOLS   R-squared (Between):              0.2536
No. Observations:                 546   R-squared (Within):               0.8642
Date:                Fri, Jul 15 2022   R-squared (Overall):              0.3443
Time:                        11:36:16   Log-likelihood                   -6004.8
Cov. Estimator:            Unadjusted                                           
                                        F-statistic:                      622.48
Entities:                          78   P-value                           0.0000
Avg Obs:                       7.0000   Distribution:                   F(5,457)
Min Obs:                       7.0000                                           
Max Obs:                       7.0000   F-statistic (robust):             622.48
                            

In [60]:
results_reg_microes_eftempo_mun.f_pooled

Pooled F-statistic
H0: Effects are zero
Statistic: 12.7270
P-value: 0.0000
Distributed: F(83,457)
WaldTestStatistic, id: 0x1434c115910

Tanto no fim do sumário quanto na linha de código acima, explicitamos os resultados do teste F para a significância dos efeitos de tempo e município, em conjunto, na regressão. Com um p-valor de 0.0, temos uma probabilidade de 0% de errar ao rejeitarmos a hipótese nula de que os efeitos em questão são estatisticamente insignificantes. Logo, com esse critério, podemos controlar para ambos os efeitos simultaneamente.

Contudo, vale destacar que individualmente o sinal do parâmetro de interesse da variável de crédito deflacionado per capita encontra-se em desacordo com o esperado pela teoria econômica, além de não ser significante nem ao nível de 10%. Nesse caso, devemos atentar para a implementação desse modelo na avaliação da política em questão.

## g)

Printando os resultados para os testes F de inclusão de efeitos fixos e organizando a tabela para comparar os coeficientes estimados pelos 3 diferentes modelos.

In [61]:
print('Teste de signicância para os efeitos fixos de município', results_reg_microes_ef.f_pooled, sep = '\n') # \n para pular uma linha
print('\nTeste de signicância para os efeitos fixos de tempo', results_reg_microes_eftempo.f_pooled, sep = '\n') 
print('\nTeste de signicância para os efeitos fixos de município e tempo', results_reg_microes_eftempo_mun.f_pooled, sep = '\n')

Teste de signicância para os efeitos fixos de município
Pooled F-statistic
H0: Effects are zero
Statistic: 12.3713
P-value: 0.0000
Distributed: F(77,463)

Teste de signicância para os efeitos fixos de tempo
Pooled F-statistic
H0: Effects are zero
Statistic: 1.0145
P-value: 0.4148
Distributed: F(6,534)

Teste de signicância para os efeitos fixos de município e tempo
Pooled F-statistic
H0: Effects are zero
Statistic: 12.7270
P-value: 0.0000
Distributed: F(83,457)


In [62]:
comp_table = pd.DataFrame({'EF_municipio' : np.round(results_reg_microes_ef.params, 3),
                           'EF_tempo' : np.round(results_reg_microes_eftempo.params, 3),
                           'EF_municipio_tempo' : np.round(results_reg_microes_eftempo_mun.params, 3)})         

comp_table

EF_municipio   EF_tempo  EF_municipio_tempo
FPM_def_pc             -12.693    -15.871               3.545
IFDM_GERAL          -78802.852  76667.334          -85612.458
cred_def_pc            114.645    -28.752              -8.575
oper_pc            -793984.398 -38735.134          169645.559
royalties_def_pc        93.203     58.524              94.871

Ao compararmos os testes F para saber se a inclusão dos efeitos fixos é adequada, verificamos que o caso em que incluímos apenas os efeitos fixos de tempo é o único não recomendado. Contudo, ao olharmos para a teoria econômica, o sinal esperado os coeficientes da variável de crédito e os testes t para a significância dessa variável nos três modelos, podemos concluir que apenas o modelo de efeitos fixos de município se torna adequado para analisar o impacto do crédito sobre o PIB das cidades comtempladas. 

Ainda, podemos dizer que o parâmetro de interesse se modifica nas estimações entre os três modelos pelo problema da variável omitida. No caso, controlar para os efeitos fixos de município parece ser a saída mais adequada, por isso, ao incluirmos apenas os efeitos fixos de tempo, o parâmetro estimado não é estatisticamente significante e não converge para o sinal esperado na teoria econômica. Na inclusão de ambos os efeitos, os efeitos fixos de tempo aparentam contaminar o resultado estimado, mesmo com a inclusão dos efeitos fixos de município, contribuindo para a incosistência e ausência de significância estatística.

O modelo com efeitos fixos de municípios aparenta produzir um estimador mais consistente, não apenas em linha com o sinal esperado pela teoria econômica, mas também em termos de significância estatística.

Por fim, podemos dizer que existem evidências, apesar de não tão consolidadas, do impacto positivo do microcrédito no PIB dos municípios.

## h)

In [63]:
reg_microes_re = plm.RandomEffects.from_formula('pib_def_pc ~ cred_def_pc + FPM_def_pc + royalties_def_pc + oper_pc + IFDM_GERAL', 
                                                data = microcred_esdata_ef)
results_reg_microes_re = reg_microes_re.fit()
print(results_reg_microes_re.summary)

                        RandomEffects Estimation Summary                        
Dep. Variable:             pib_def_pc   R-squared:                        0.8626
Estimator:              RandomEffects   R-squared (Between):              0.9389
No. Observations:                 546   R-squared (Within):               0.8001
Date:                Fri, Jul 15 2022   R-squared (Overall):              0.9183
Time:                        11:36:18   Log-likelihood                   -6198.2
Cov. Estimator:            Unadjusted                                           
                                        F-statistic:                      679.33
Entities:                          78   P-value                           0.0000
Avg Obs:                       7.0000   Distribution:                   F(5,541)
Min Obs:                       7.0000                                           
Max Obs:                       7.0000   F-statistic (robust):             679.33
                            

Teste de Huasman, implementado manualmente de acordo com HEISS, F.. Using Python for Introductory Econometrics, 2020.

> Link para acesso aos exemplos do livro citado: http://www.upfie.net/index.html

In [64]:
import scipy.stats as stats


# (I) Encontrando os coeficientes em comum e as covariâncias:
common_coef = set(results_reg_microes_ef.params.index).intersection(results_reg_microes_re.params.index)
b_ef_cov = results_reg_microes_ef.cov
b_re_cov = results_reg_microes_re.cov

# (II) Calculando as diferenças entre EF e EA:
b_diff = np.array(results_reg_microes_ef.params[common_coef] - results_reg_microes_re.params[common_coef])
df = len(b_diff)
b_diff.reshape((df, 1))
b_cov_diff = np.array(b_ef_cov.loc[common_coef, common_coef] -
                      b_re_cov.loc[common_coef, common_coef])
b_cov_diff.reshape((df, df))

# (III) Calculando a estatística de teste e o pvalor:
stat = abs(np.transpose(b_diff) @ np.linalg.inv(b_cov_diff) @ b_diff)
pval = 1 - stats.chi2.cdf(stat, df)

print(f'stat: {stat}\n')
print(f'pval: {pval}\n')

stat: 455.40167733857595

pval: 0.0



Lembrando que:

$H_0$: $q = 0$, $\hat{\beta_{EF}} = \hat{\beta_{EA}}$

$H_1$: $q \neq 0$, $\hat{\beta_{EF}} \neq \hat{\beta_{EA}}$

Ao comparar os parâmetros dos modelos de efeitos fixos para municípios com o de efeitos aleatórios, podemos dizer que há diferenças estatisticamente significantes, dado o p-valor = 0.0, que indica que temos 0% de chance de estarmos equivocados ao rejeitar $H_0$. 

Nesse caso, e reforçando os argumentos ao longo da questão, podemos concluir que, pela significância estatística do parâmetro, o sinal de acordo com o esperado com a teoria econômica e a significância para os efeitos fixos, o modelo de efeitos fixos para municípios é o mais adequado para o problema.

# Questão 5)

In [65]:
mercado_peixe = pd.read_excel('mercado_peixe.xlsx')
mercado_peixe

LogPrice  Stormy  LogQuantity
0   -0.430783       1     8.994421
1    0.000000       1     7.707063
2    0.072321       0     8.350194
3    0.247139       1     8.656955
4    0.664327       1     7.844241
..        ...     ...          ...
106 -0.798508       0     8.610683
107 -0.087011       0     7.162397
108  0.184922       0     7.362010
109  0.223143       0     8.764053
110  0.561118       0     8.328451

[111 rows x 3 columns]

## a) 

Apenas a curva de demanda pode ser estimada. Isso porque apenas a curva de oferta possui uma variável exógena (Stormy), além da variável endógena (LogPrice), que pode ser utilizada como instrumento para estimar consistentemente a equação de demanda. A curva de demanda, por sua vez, possui apenas a variável endógena (LogPrice), logo, não possui temos um instrumento que pode ser utilizado para estimar a equação de oferta.

## b)

O sistema segue abaixo:

$
\begin{cases}
log(Q_s) = \beta_0 + \alpha_{11}log(price) + \beta_1stormy + u_1\\
log(Q_d) = \beta_0 + \alpha_{12}log(price) + u_2
\end{cases}
$

## c)

In [66]:
reg_reduz = smf.ols('LogPrice ~ Stormy', data = mercado_peixe)
results_reg_reduz = reg_reduz.fit()
print(results_reg_reduz.summary())

                            OLS Regression Results                            
Dep. Variable:               LogPrice   R-squared:                       0.160
Model:                            OLS   Adj. R-squared:                  0.152
Method:                 Least Squares   F-statistic:                     20.69
Date:                Fri, 15 Jul 2022   Prob (F-statistic):           1.41e-05
Time:                        11:36:19   Log-Likelihood:                -40.516
No. Observations:                 111   AIC:                             85.03
Df Residuals:                     109   BIC:                             90.45
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -0.2903      0.040     -7.336      0.0

Com base na nossa regra de bolso, o instrumento Stormy é forte, dado que temos uma estatística |t| = 4.549 > 3.2 e uma estatística |F| = 20.69 > 10. 

## d)

In [67]:
import linearmodels.iv as iv

In [68]:
reg_demanda_peixe = iv.IV2SLS.from_formula('LogQuantity ~ 1 + [LogPrice ~ Stormy]', data = mercado_peixe)
results_reg_demanda_peixe = reg_demanda_peixe.fit(cov_type = 'unadjusted', debiased = True)
print(results_reg_demanda_peixe.summary)

                          IV-2SLS Estimation Summary                          
Dep. Variable:            LogQuantity   R-squared:                     -0.0002
Estimator:                    IV-2SLS   Adj. R-squared:                -0.0094
No. Observations:                 111   F-statistic:                    5.4017
Date:                Fri, Jul 15 2022   P-value (F-stat)                0.0220
Time:                        11:36:20   Distribution:                 F(1,109)
Cov. Estimator:            unadjusted                                         
                                                                              
                             Parameter Estimates                              
            Parameter  Std. Err.     T-stat    P-value    Lower CI    Upper CI
------------------------------------------------------------------------------
Intercept      8.3138     0.1146     72.532     0.0000      8.0866      8.5410
LogPrice      -1.0824     0.4657    -2.3242     0.02

Com o coeficiente de LogPrice = -1.08, podemos considerar que estamos diante de uma curva de demanda com elasticidade unitária, ou seja, uma varição de 1% no preço, *ceteris paribus*, teremos uma variação no sentido oposto na quantidade demandada. 

# Questão 6)

In [69]:
imoveis = pd.read_csv('imoveis.csv', sep = ';')
imoveis

ano  dorm  banho  garag  elev  unid_andar  andares  area_util   preco  \
0     1985     4      3      3     2           1       17        277  368480   
1     1985     4      3      3     2           1       21        248  382220   
2     1985     4      4      4     2           1       17        245  534000   
3     1985     4      2      2     3           4       15        113  214930   
4     1985     2      1      1     2           4       13         81  127410   
...    ...   ...    ...    ...   ...         ...      ...        ...     ...   
2509  1991     2      1      1    12           4       12         50  125370   
2510  1991     3      1      1     2           4        7         54  187770   
2511  1991     2      1      1     2           8        8         43  141040   
2512  1991     2      1      1     2           4        9         56  139390   
2513  1991     2      1      1     3           8       15         50  132360   

      tratado  
0           1  
1           1  
2           1  
3           1  
4           1  
...       ...  
2509        0  
2510        0  
2511        0  
2512        0  
2513        0  

[2514 rows x 10 columns]

Criando as dummies para os anos pré e pós tratamento e a coluna de interação entre as variáveis dos anos de pré e pós tratamento com a variável dummy para tratados:

In [70]:
# (I) Criando a variável ano para os períodos pré e pós-tratamento

conditions = [
    (imoveis['ano'] <= 1988),
    (imoveis['ano'] > 1988)]

choices = [0, 1]
imoveis['ano_tratamento'] = np.select(conditions, choices)

# (II) Criando a variável de interação

imoveis['ano_tratamento_tratado'] = imoveis['ano_tratamento'] * imoveis['tratado'] 
imoveis

ano  dorm  banho  garag  elev  unid_andar  andares  area_util   preco  \
0     1985     4      3      3     2           1       17        277  368480   
1     1985     4      3      3     2           1       21        248  382220   
2     1985     4      4      4     2           1       17        245  534000   
3     1985     4      2      2     3           4       15        113  214930   
4     1985     2      1      1     2           4       13         81  127410   
...    ...   ...    ...    ...   ...         ...      ...        ...     ...   
2509  1991     2      1      1    12           4       12         50  125370   
2510  1991     3      1      1     2           4        7         54  187770   
2511  1991     2      1      1     2           8        8         43  141040   
2512  1991     2      1      1     2           4        9         56  139390   
2513  1991     2      1      1     3           8       15         50  132360   

      tratado  ano_tratamento  ano_tratamento_tratado  
0           1               0                       0  
1           1               0                       0  
2           1               0                       0  
3           1               0                       0  
4           1               0                       0  
...       ...             ...                     ...  
2509        0               1                       0  
2510        0               1                       0  
2511        0               1                       0  
2512        0               1                       0  
2513        0               1                       0  

[2514 rows x 12 columns]

## a) 

In [71]:
reg_did_imoveis = smf.ols('preco ~ tratado + ano_tratamento + ano_tratamento_tratado', 
                           data = imoveis)
results_reg_did_imoveis = reg_did_imoveis.fit()
print(results_reg_did_imoveis.summary())
print('\nParâmetro Diff-in-Diff:\n', round(results_reg_did_imoveis.params[3], 2))        

                            OLS Regression Results                            
Dep. Variable:                  preco   R-squared:                       0.004
Model:                            OLS   Adj. R-squared:                  0.003
Method:                 Least Squares   F-statistic:                     3.308
Date:                Fri, 15 Jul 2022   Prob (F-statistic):             0.0194
Time:                        11:36:21   Log-Likelihood:                -36178.
No. Observations:                2514   AIC:                         7.236e+04
Df Residuals:                    2510   BIC:                         7.239e+04
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                             coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------
Intercept               4.16

Em um modelo simples, considerando apenas os anos como variáveis explicativas para avaliar o impacto da construção do metro nos preços dos imóveis antes de depois da construção, temos que um imóvel na área tratada teve seu preço valorizado em R$ 163,176.47 em média. Ou seja, o impacto da construção do metrô foi positivo em relação aos preços dos imóveis.

## b)

In [72]:
reg_did_imoveis_controls = smf.ols('preco ~ dorm + banho + garag + elev + unid_andar + andares +' 
                                    'area_util + tratado + ano_tratamento + ano_tratamento_tratado', 
                                    data = imoveis)
results_reg_did_imoveis_controls = reg_did_imoveis_controls.fit()
print(results_reg_did_imoveis_controls.summary())
print('\nParâmetro Diff-in-Diff:\n', round(results_reg_did_imoveis_controls.params[10], 2))

                            OLS Regression Results                            
Dep. Variable:                  preco   R-squared:                       0.653
Model:                            OLS   Adj. R-squared:                  0.652
Method:                 Least Squares   F-statistic:                     472.0
Date:                Fri, 15 Jul 2022   Prob (F-statistic):               0.00
Time:                        11:36:21   Log-Likelihood:                -34851.
No. Observations:                2514   AIC:                         6.972e+04
Df Residuals:                    2503   BIC:                         6.979e+04
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
                             coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------
Intercept              -1.20

Ao incluirmos as demais variáveis de controle do banco de dados, podemos reparar que o parâmetro de Diferenças em Diferenças estimado muda de valor, reduzindo-se. Ao aplicar os testes estatísticos, podemos ver que individualmente, exceto as variáveis *tratado* e *elev*, todas as variáveis do modelo são estatísticamente significantes ao nível de siginificância de 5%, a partir do teste t e conferindo os p-valores. Do ponto de vista conjunto, por meio do teste F, temos que todos os parâmetros são significativos.

Assim, podemos visualizar que com os controles, o impacto da construção do metro nos preços dos imóveis em regiões próximas (tratados) foi de uma valorização, em termos de preço, de R$ 84,759.63.